download/download.go

```diff
@@ -72,35 +73,68 @@ func (d Downloader) Download(ctx context.Context, pair, timeframe string, output
 
 	parameters.Start = time.Date(parameters.Start.Year(), parameters.Start.Month(), parameters.Start.Day(),
 		0, 0, 0, 0, time.UTC)
-	parameters.End = time.Date(parameters.End.Year(), parameters.End.Month(), parameters.End.Day(),
-		0, 0, 0, 0, time.UTC)
+
+	if now.Sub(parameters.End) > 0 {
+		parameters.End = time.Date(parameters.End.Year(), parameters.End.Month(), parameters.End.Day(),
+			0, 0, 0, 0, time.UTC)
+	} else {
+		parameters.End = now
+	}
 
 	candlesCount, interval, err := candlesCount(parameters.Start, parameters.End, timeframe)
 	if err != nil {
 		return err
 	}
+	candlesCount++
 
 	log.Infof("Downloading %d candles of %s for %s", candlesCount, timeframe, pair)
 	info := d.exchange.AssetsInfo(pair)
 	writer := csv.NewWriter(recordFile)
+
+	progressBar := progressbar.Default(int64(candlesCount))
+	lostData := 0
+	isLastLoop := false
+
 	for begin := parameters.Start; begin.Before(parameters.End); begin = begin.Add(interval * batchSize) {
 		end := begin.Add(interval * batchSize)
 		if end.After(parameters.End) {
 			end = parameters.End
+			isLastLoop = true
+		} else {
+			end = end.Add(-1 * time.Second)
 		}
 
 		candles, err := d.exchange.CandlesByPeriod(ctx, pair, timeframe, begin, end)
 		if err != nil {
 			return err
 		}
 
+		count := 0
 		for _, candle := range candles {
 			err := writer.Write(candle.ToSlice(int(info.PriceDecimalPrecision)))
 			if err != nil {
 				return err
 			}
+			count++
+		}
+
+		if !isLastLoop {
+			lostData += batchSize - count
 		}
+
+		if err = progressBar.Add(count); err != nil {
+			log.Warningf("update progresbar fail: %s", err.Error())
+		}
+	}
+
+	if err = progressBar.Close(); err != nil {
+		log.Warningf("close progresbar fail: %s", err.Error())
 	}
+
+	if lostData > 0 {
+		log.Warningf("found lost data %d rows", lostData)
```

Comment:
A suggestion, What do you think?
```suggestion
		log.Warningf("%d missing candles", lostData)
```

Reply:
it's better than my message xD